### PR TITLE
feat(cache): Expired users are now cleaned from the database

### PR DESCRIPTION
**Pull request by denisonbarbosa**
User entries that exceed the offline expiration period will be removed from the database unless they are in an active session. Groups that end up empty after the user removal will also be removed.

UDENG-1417

### CODE REVIEW

File: internal/cache/db_test.go
Comment (denisonbarbosa) [inline internal/cache/db_test.go:513]: I'm pretty sure this is not the best solution for the "noise" problem, but I wasn't sure what was the best way to handle the amount of diff that would be generated when updating most of the test cases and fixtures.

@didrocks, any insights on this? Maybe it should be tackled on a separate `tests(cache)` PR after this one gets merged? I'll keep it as a draft until we decide how to handle this.

Comment (didrocks) [reply internal/cache/db_test.go:513]: Maybe instead of this fixture, we can have an exported_tests on cache which set a user last login back to the path? That way, we open the cache, all is good, and we reset the time, and test triggering the cleaning, wdyt?

File: internal/cache/serialization.go
Comment (didrocks) [reply internal/cache/serialization.go:130]: I didn’t look far, but I feel the logic is incorrect anyway. We don’t care about the CURRENT_USER. It will be root anyway in production for the service, we care about any logged in users having some processes opened.

File: internal/cache/db_test.go
Comment (denisonbarbosa) [reply internal/cache/db_test.go:513]: I'll try this out. Thanks!

File: internal/cache/serialization.go
Comment (denisonbarbosa) [reply internal/cache/serialization.go:130]: So we should use something like "ACTIVE_USER" and fetch one from `utmp`?

File: internal/cache/db_test.go
Comment (didrocks) [reply internal/cache/db_test.go:513]: btw, most of the tests should not trigger cleaning, hence a With function by default in the tests to put the cleaning time before 2004.

File: internal/cache/serialization.go
Comment (didrocks) [reply internal/cache/serialization.go:130]: The issue from `utmp` is that it’s not that reliable. We did with jibel multiple attempts and it’s not clear.

If you `ls -ld /proc/*`, you will see for the pids the user owning the process. So we can probably loop over that dir

File: internal/services/pam/testdata/TestIsAuthenticated/golden/update_existing_db_on_success/cache.db
Comment (didrocks) [inline internal/services/pam/testdata/TestIsAuthenticated/golden/update_existing_db_on_success/cache.db:13]: I guess you should avoid purging for this one too.

**Review by didrocks — CHANGES_REQUESTED** [review]
no long general comment, we already discussed most of them online :)

File: internal/cache/delete.go
Comment (denisonbarbosa) [reply internal/cache/delete.go:32]: Since these are private functions, shouldn't we return the error and leave it up to the caller to decide whether it should panic or just log it?

Comment (didrocks) [reply internal/cache/delete.go]: I don’t think so, this is an invalid use of the function, a programming error, so I would say let’s panic where the panic occurs from, unconditionnally.

File: internal/cache/delete.go
Comment (didrocks) [inline internal/cache/delete.go:34]: really, tag it as "programming error", so that we know we are not going to send this to the finale user.

Comment (didrocks) [reply internal/cache/delete.go:34]: And you removed the comment to: which stilll explains the panic I guess.

File: internal/cache/db_test.go
Comment (didrocks) [inline internal/cache/db_test.go:97]: This is not what I meant on my previous comment.
Set a default expiration, then, if tc.expirationDate != "", compute the new expiration,. and after closing the if, do:
`cacheOpts := []cache.Option{cache.WithExpirationDate(expiration)}`

File: internal/cache/db_test.go
Comment (didrocks) [reply internal/cache/db_test.go]: This was this comment explaining it :)

File: internal/cache/db.go
Comment (didrocks) [inline internal/cache/db.go:283]: there is still an online here

File: internal/cache/db.go
Comment (didrocks) [inline internal/cache/db.go:301]: I would still capture an error for the day we change the implementation internal, it doesn’t bring us anything to ignore it, no?

File: internal/cache/db.go
Comment (didrocks) [inline internal/cache/db.go:332]: sounds a little bit forced, would there be another way of having cleaning failing without exposing the procdir?

File: internal/cache/db_test.go
Comment (didrocks) [inline internal/cache/db_test.go:42]: nitpick: I would rename the dbFile to "only_old_users"

**Review by didrocks — CHANGES_REQUESTED** [review]
Some comments were probably not clear enough, I clarified them!

I think this is going into the right direction.

However, the returned error is not covered here and I think we can cover with a corrupted database:
We try to delete an user is part of a group which doesn’t have an entry for this group, if that makes sense.

```
        for _, gid := range userToGroups.GIDs {
                if err := deleteUserFromGroup(buckets, uid, gid); err != nil {
                        return err
                }
        }
```

Does it make sense? Also, if we can not expose /proc directory (as we are not really using it for real fixtures, if it did, I would be in favor of it, but apart from running the tests as root, we can’t chown to another user the directories)

File: internal/cache/db.go
Comment (denisonbarbosa) [reply internal/cache/db.go:301]: We'll lose some coverage %, but it's fine I think.

File: internal/cache/db_test.go
Comment (denisonbarbosa) [reply internal/cache/db_test.go:97]: Oops, sorry... I really misunderstood this one completely.

File: internal/cache/delete.go
Comment (denisonbarbosa) [reply internal/cache/delete.go:34]: As in `PROGRAMMING ERROR: could not...` or as in `CRITICAL: programming error: could not...`?

File: internal/cache/db.go
Comment (denisonbarbosa) [reply internal/cache/db.go:332]: I don't think so. The only points of failure in the cleanup function are:
```Go
activeUsers, err := getActiveUsers(procDir)
if err != nil {
	return err
}

buckets, err := getAllBuckets(tx)
if err != nil {
	return err
}
```

We could fail on getAllBuckets, but this would imply a corrupted database and there's no way we can get this in the first cleanup call (when creating the cache object).

File: internal/cache/delete.go
Comment (didrocks) [reply internal/cache/delete.go:34]: I would say do like the rest of the code

File: internal/cache/db.go
Comment (didrocks) [reply internal/cache/db.go:332]: If you look at your coverage, it seems you triggered this error:
```
                info, err := dirEntry.Info()
                if err != nil {
                        return nil, err
                }
```

File: internal/cache/db_test.go
Comment (didrocks) [inline internal/cache/db_test.go:98]: nitpick: I would remove the blank line before, and then add a blank line for other options (basically, the option belongs to what we computed above IMHO rather than grouping options)

**Pull request merged**
